### PR TITLE
feat(sec-core): add sqldb writer&reader and query command at cli interface

### PR DESCRIPTION
**Pull request by RemindD**
## Description

Core storage engine:
- Add SqliteEventWriter: thread-safe, corruption-resilient, self-pruning SQLite writer with convergent schema migration and dual-write support
- Add SqliteEventReader: sqlite3.Row name-based access, defense-in-depth SQL injection prevention (allowlist + regex assertion in count_by)
- Introduce `result` field (succeeded/failed) in SecurityEvent schema, replacing the _error suffix convention in event_type

CLI `events` command:
- kubectl-style table output with dynamic column widths (default format)
- Support --output table|json|jsonl for progressive disclosure
- Filters: --event-type, --category, --trace-id, --since, --until, --last-hours
- Aggregation: --count, --count-by (category|event_type|trace_id)
- Pagination: --limit, --offset
- Soft validation: warn on unknown event_type/category without rejecting
- Timestamp truncation to seconds in table display for readability
- Inline usage examples in source-level section comments

Event model refactoring:
- lifecycle.on_error now emits event_type=`action` with result="failed" instead of event_type=`action`_error (cleaner query semantics)
- SummaryBackend upgraded from stub to functional aggregation engine (removed db_health diagnostic logic as unnecessary)

Test infrastructure:
- Function-scoped e2e fixture (per-test DB isolation, no ordering dependency)
- Session-scoped unit-test conftest redirects AGENT_SEC_DATA_DIR to tmp
- Comprehensive e2e coverage: event logging, count, count-by, filters, table/json/jsonl output, CLI validation error paths
- Unit tests for writer (multiprocess, corruption recovery, pruning), reader (query, count_by, edge cases), schema, and summary backend

## Related Issue

closes #

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional change)
- [ ] Performance improvement
- [ ] CI/CD or build changes

## Scope

- [ ] `cosh` (copilot-shell)
- [x] `sec-core` (agent-sec-core)
- [ ] `skill` (os-skills)
- [ ] `sight` (agentsight)
- [ ] Multiple / Project-wide

## Checklist

- [ ] I have read the [Contributing Guide](../CONTRIBUTING.md)
- [ ] My code follows the project's code style
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the documentation accordingly
- [ ] For `cosh`: Lint passes, type check passes, and tests pass
- [ ] For `sec-core` (Rust): `cargo clippy -- -D warnings` and `cargo fmt --check` pass
- [ ] For `sec-core` (Python): Ruff format and pytest pass
- [ ] For `skill`: Skill directory structure is valid and shell scripts pass syntax check
- [ ] For `sight`: `cargo clippy -- -D warnings` and `cargo fmt --check` pass
- [ ] Lock files are up to date (`package-lock.json` / `Cargo.lock`)

### CODE REVIEW

**Comment by yangdao479** [comment]
请参考开发手册（DEVELOPMENT.md） 
1. 不需要使用from __future__ import annotations，python3.11不需要
2. 不要在函数内进行import（PEP8）

File: src/agent-sec-core/agent-sec-cli/src/agent_sec_cli/security_events/sqlite_writer.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        # Introspect existing columns for convergent migration
+        existing = {
+            row[1] for row in conn.execute("PRAGMA table_info(security_events)")
+        }
+        for col, typedef in _COLUMNS.items():
+            if col not in existing:
+                assert re.match(r"^[a-z_]+$", col)  # noqa: S101
+                conn.execute(f"ALTER TABLE security_events ADD COLUMN {col} {typedef}")
+
+        # Version-gated escape hatch
```
Comment (edonyzpc) [inline src/agent-sec-core/agent-sec-cli/src/agent_sec_cli/security_events/sqlite_writer.py:249]: WAL/SHM 文件清理依赖 `with_suffix()` 替换扩展名，如果 DB 文件名不以 `.db` 结尾（如裸文件名 `security-events`），会生成错误路径。

建议改为字符串拼接：
```python
wal = Path(str(self._path) + "-wal")
shm = Path(str(self._path) + "-shm")
```
这与 SQLite WAL 文件的实际命名规则完全一致。

File: src/agent-sec-core/agent-sec-cli/src/agent_sec_cli/security_events/sqlite_reader.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        until : str, optional
+            Exclusive upper bound (ISO-8601 timestamp).
+
+        Returns
+        -------
+        dict[str, int]
+            Mapping of field value to event count.
+
+        Raises
+        ------
```
Comment (edonyzpc) [inline src/agent-sec-core/agent-sec-cli/src/agent_sec_cli/security_events/sqlite_reader.py:222]: `assert` 在 Python 的 `-O` 优化模式下会被跳过（`python -O`），如果未来有代码误删或绕过上方的 allowlist 检查，这里的 assert 不会触发。

建议改为显式 raise：
```python
if not re.match(r"^[a-z_]+$", group_field):
    raise ValueError(f"group_field contains invalid characters: {group_field!r}")
```
防御逻辑在优化模式下同样生效。

**Review by edonyzpc — CHANGES_REQUESTED** [review]
## Code Review Summary

PR 整体设计清晰、工程质量良好：双写模式、WAL 模式、corruption 自愈、SQL 注入防御、fire-and-forget 都有合理的考量。以下列出需要修复的阻断问题和建议改进项。

---

### 🔴 阻断问题（Blocking Issues）

#### 1. `cli.py` — `--count --last-hours` 将全量事件加载到内存再计数，效率低下

当前 count 模式中对 `--last-hours` 的处理：

```python
# cli.py — count mode
if last_hours is not None:
    result = reader.query_last_hours(
        last_hours, event_type=event_type, category=category
    )
    typer.echo(json.dumps(len(result), ensure_ascii=False, indent=2))  # ← 全量加载
```

`query_last_hours()` 内部有硬编码 `LIMIT 10000`，这意味着当事件量超过 1 万时，`--count` 返回的结果将是错误的（最多为 10000），而不是实际总数。

应直接使用 `reader.count()` 并传入 ISO 时间戳：

```python
if count:
    resolved_since, resolved_until = _resolve_time_range(last_hours, since, until)
    result = reader.count(
        event_type=event_type,
        category=category,
        since=resolved_since,
        until=resolved_until,
    )
    typer.echo(json.dumps(result, ensure_ascii=False, indent=2))
    raise typer.Exit(code=0)
```

`_resolve_time_range()` 已经统一处理了 `--last-hours` → ISO 转换，直接复用即可消除这个不一致性。

---

### 🟡 应修复问题（Should Fix）

#### 2. `config.py` — `mkdir(parents=True, mode=0o700)` 仅保护叶目录，中间目录权限未收紧

```python
primary_dir.mkdir(parents=True, exist_ok=True, mode=0o700)
```

`Path.mkdir(parents=True, mode=...)` 的语义是：**只有最终叶目录**使用指定 mode，`parents=True` 创建的中间目录使用系统默认 umask。这意味着如果 `/var/log/agent-sec/` 不存在，`/var/log` 之类的中间目录可能以更宽松的权限创建。

建议在创建后明确 `chmod`，或者参考已有的 `_safe_tmp_dir()` 模式对每层目录进行 lstat 校验。

#### 3. `tests/unit-test/conftest.py` — session 级别 fixture 未重置模块级单例，可能引发测试污染

```python
@pytest.fixture(autouse=True, scope="session")
def _isolate_data_dir(tmp_path_factory):
    data_dir = tmp_path_factory.mktemp("unit-test-data")
    os.environ["AGENT_SEC_DATA_DIR"] = str(data_dir)
    yield
    os.environ.pop("AGENT_SEC_DATA_DIR", None)
```

如果在 fixture 执行前（模块导入阶段）某个测试文件已触发了 `get_sqlite_writer()` 或 `get_reader()` 的懒加载，单例会指向旧路径。建议在 `yield` 前主动重置：

```python
import agent_sec_cli.security_events as _ev
os.environ["AGENT_SEC_DATA_DIR"] = str(data_dir)
_ev._writer = None
_ev._sqlite_writer = None
_ev._reader = None
yield
_ev._writer = None
_ev._sqlite_writer = None
_ev._reader = None
os.environ.pop("AGENT_SEC_DATA_DIR", None)
```

#### 4. `tests/unit-test/security_middleware/test_lifecycle.py` — 集成测试混入单元测试，且有单例泄漏风险

`test_error_event_writes_to_sqlite` 做了真实 SQLite I/O、环境变量切换和模块状态操作，属于集成测试范畴，放在 unit-test 目录下会增加 unit test 的时间和依赖。

更严重的是：测试末尾调用了 `get_sqlite_writer().close()`，但没有将 `_events_mod._sqlite_writer` 重置回 `None`。后续代码若调用 `get_sqlite_writer()` 会返回一个 `_conn=None` 的已关闭实例，虽然 `_ensure_connection()` 会尝试重连，但重连目标路径已经指向该测试的临时目录（环境变量已在 finally 中恢复）。

建议将该测试移至 `tests/e2e/cli/` 目录，或在 `finally` 块中同步重置 `_events_mod._sqlite_writer = None`。

---

### 🔵 次要建议（Minor Suggestions）

#### 5. `sqlite_reader.py` — `assert` 在 `-O` 优化模式下可被跳过

```python
# sqlite_reader.py — count_by()
assert re.match(r"^[a-z_]+$", group_field), f"..."
```

注释中已解释这是 belt-and-suspenders，allowlist 是主防线。但为了避免误导，建议改为显式 raise：

```python
if not re.match(r"^[a-z_]+$", group_field):
    raise ValueError(f"group_field contains invalid characters: {group_field!r}")
```

这样无论是否使用 `-O` 优化都能保证行为一致。

#### 6. `sqlite_writer.py` — `_handle_corruption()` 中 WAL/SHM 文件的清理逻辑依赖文件扩展名约定

```python
wal = self._path.with_suffix(".db-wal")
shm = self._path.with_suffix(".db-shm")
```

`Path.with_suffix()` 会替换最后一个后缀。如果 DB 路径不是 `.db` 结尾（如 `security-events`，无扩展名），这里会产生错误的文件名。建议用字符串拼接保持健壮性：

```python
wal = Path(str(self._path) + "-wal")
shm = Path(str(self._path) + "-shm")
```

这与 SQLite WAL 文件的实际命名规则（`{dbfile}-wal`）完全一致。

#### 7. `cli.py` — `_VALID_EVENT_TYPES` 和 `_VALID_CATEGORIES` 需要手动维护

注释说明这些常量需要与 `lifecycle.py` 保持同步，但依赖人工维护。建议在注释中增加一个明确的 TODO，或者从 `lifecycle.py._ACTION_CATEGORY` 动态派生，以防未来添加新 action 时遗漏更新。

#### 8. `sqlite_reader.py::query_last_hours()` — `_build_where()` 未复用

`query()` 通过 `_build_where()` 构建 WHERE 子句，而 `query_last_hours()` 内联重复了类似的逻辑。虽然有性能上的理由（避免 epoch → ISO → epoch 转换），但这造成了代码重复。可以考虑给 `_build_where()` 增加一个 `since_epoch` 参数，统一两个路径。

---

### ✅ 值得认可的设计

- **双写容错模式**：JSONL 和 SQLite 互相独立，任一失败不影响另一个，fire-and-forget 语义统一。
- **WAL + `busy_timeout`**：多进程写入场景的正确选择，有效避免 `SQLITE_LOCKED`。
- **Corruption 自愈**：检测到损坏时删除 DB（连同 WAL/SHM）并重建，不丢当次事件。
- **SQL 注入防御**：`count_by()` allowlist + 正则 double-check，防御设计清晰。
- **测试隔离**：E2E 使用 function-scoped fixture，单元测试 session-scoped 隔离，思路正确。
- **事件模型重构**：用 `result=failed` 替换 `_error` 后缀，使得按类型查询不再需要处理两种命名约定，更符合 CQRS 思想。

---

**结论**：Issue 1（`--count --last-hours` 计数错误）是功能性 bug，会在高负载场景下返回错误结果，必须修复。其余问题建议在合并前评估。

**Comment by RemindD** [comment]
@edonyzpc 最新commit中已修复以上问题。
1. 移除了query_last_hours，统一使用reader.query()
2. 创建后明确 chmod
3. fixture 重置sql config
4. 已移到e2e test下
5. 两处assert 均改为raise
6. 已修复
7. 改为从lifecycle.py 动态导入
8. query_last_hours已移除

**Review by edonyzpc — APPROVED** [review]
LGTM
